src/MDMTransaction.m

```diff
@@ -30,17 +32,58 @@ - (instancetype)init {
 }
 
 - (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target {
-  MDMTransactionLog *log = [MDMTransactionLog new];
-  log.plans = @[ [plan copy] ];
-  log.target = target;
+  [self commonAddPlan:plan toTarget:target withName:nil];
+}
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target named:(NSString *)name {
+  if (name.length && [self planForName:name withTarget:target]) {
+    [self commonRemovePlan:plan fromTarget:target named:name];
+  }
+  [self commonAddPlan:plan toTarget:target withName:name];
+}
+
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  [self commonRemovePlan:[self planForName:name withTarget:target] fromTarget:target named:name];
+}
+
+- (void)commonAddPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:[plan copy] target:target name:name removal:FALSE];
+  [_logs addObject:log];
+}
+
+- (void)commonRemovePlan:(NSObject<MDMPlan> *)plan fromTarget:(nonnull id)target named:(NSString *)named {
+  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:plan target:target name:named removal:TRUE];
   [_logs addObject:log];
 }
 
 - (NSArray<MDMTransactionLog *> *)logs {
   return _logs;
 }
 
+- (id<MDMPlan>)planForName:(NSString *)name withTarget:(id)target {
```

Comment:
- [ ] Delete this method. It will encourage too much state manipulation within the transaction object.